### ISSUE TITLE
BPF: implementing XDP bit TC filter

### ISSUE THREAD

**Issue by MiniPierre** [PowerDNS/pdns#10451]
- Program: dnsdist
 - Issue type: Feature request

### Short description
It is actually more an info request than a feature one. I'm currently working on implementing the Trucate BPF filter discussed in this issue: https://github.com/PowerDNS/pdns/issues/10324

### Usecase
N/A

### Description
I am trying to implement another BPF filter using XDP to allow sending DNS answer with the TC bit set. Having never worked with this technology, I tried to modify the bpf-filter.ebpf.src file to do some test but I cannot manage to compile it, as it is never imported in any other file (only the compiled files are used). Could you please tell me which tool you used ?

**Comment by rgacogne** [comment]
Hi! Great news, I would very much like to have that feature :) That file is indeed not directly compiled, we ship a lightly modified version of the eBPF code generated from that file, to prevent having to install all the tooling required to do the compilation.
I would suggest trying to get some code working using bcc first, especially since there is now an example quite close to what we do in the bcc repo: https://github.com/iovisor/bcc/blob/master/examples/networking/dns_matching/dns_matching.c
Once you have some code that can return an answer with the TC bit set I will gladly help you integrate that inside dnsdist.

**Comment by rgacogne** [comment]
You can see how that code is dynamically compiled and used by looking at https://github.com/iovisor/bcc/blob/master/examples/networking/dns_matching/dns_matching.py, by the way.

**Comment by MiniPierre** [comment]
Thank you for all those information ! I will have a close look at them
It will actually take quite a long time as I am not used to C++ but be sure I will keep you informed when I will have something working

**Comment by rgacogne** [comment]
I just realized that this example, and what we are currently doing, uses eBPF socket filters which (as far as I can tell) can't craft an answer directly. For that you indeed need XDP, as you mentioned, so it's a bit different. You could start by having a look at https://blog.nlnetlabs.nl/journeying-into-xdp-part-1-augmenting-dns/, which takes you through most of what you need to know to write the eBPF code generating a TC=1 answer from a query. Then we can figure out how to integrate into dnsdist together.

**Comment by MiniPierre** [comment]
Working on it a little more, I realized that it could be possible to dynamically pass the network port to filter by using a BPF map storing the network port. By doing so, the XDP filter will check the port to filter for every packet, but I do not think it would use a lot of processing ressources

**Comment by rgacogne** [comment]
I'm pretty sure we can live with filtering on port 53 first, and figure out later if/how we can fix that :-) As long as dnsdist can populate the BPF maps used by the XDP program to do source IP/qname filtering (and I see no reason why it would not) I'd even be happy with the XDP program being completely outside of dnsdist if needed.

**Comment by RobinGeuze** [comment]
What I've seen in other applications that use eBPF (Project Calico in this case) is that they insert a place holder value that is most likely to be unique and then do a hard string replace before loading the program into the kernel. See [here](https://github.com/projectcalico/felix/blob/2884c1e48f3db0b38b73407698601b54a19a01d2/bpf-gpl/bpf.h#L221) and [here](https://github.com/projectcalico/felix/blob/2884c1e48f3db0b38b73407698601b54a19a01d2/bpf/binary.go#L127) (its golang, but the general idea should be the same.

**Comment by MiniPierre** [comment]
Just to keep you updated, the feature is currently under development, I can now filter with the TC bit for IPv4 and IPv6 addresses but not yet for QName (it is actually working but the compiler complains about the filter having too much instructions, have to optimize things a bit)
I will also have to rewrite the bpf-filter.cc file to make it work with new adjustements
I am also thinking of rewriting the current BPF blocking filter with XDP, which would only require to replace the "set TC bit and resend the packet" part with a simple XDP_DROP action to harmonize things and make the filter faster

**Comment by rgacogne** [comment]
Wonderful!

> it is actually working but the compiler complains about the filter having too much instructions, have to optimize things a bit

Yeah, it's painful. Sub-program can help, but only if you reach the program limit and not the global one.

> I will also have to rewrite the bpf-filter.cc file to make it work with new adjustements

What would the changes be? I wouldn't worry too much about that, I can take care part of that part later.

> I am also thinking of rewriting the current BPF blocking filter with XDP, which would only require to replace the "set TC bit and resend the packet" part with a simple XDP_DROP action to harmonize things and make the filter faster

That would be very nice as long as we don't break existing setups, otherwise I'd prefer having both options available for a while, deprecating the existing one then removing it after a few versions.

**Comment by MiniPierre** [comment]
> What would the changes be? I wouldn't worry too much about that, I can take care part of that part later.
The biggest one is the fact that you are storing QNames in BPF Table like that (counter, qtype) -> qname, which I did not understood, I was thinking of replacing it by (qname, qtype) -> counter, except if it does break things. It also need to receive the interfaces on which the filter must be attached as a parameter, and change the filter type

> That would be very nice as long as we don't break existing setups, otherwise I'd prefer having both options available for a while, deprecating the existing one then removing it after a few versions.
That might be a better idea to have it in parallel, some users might prefer to have a filter listening on the socket rather than specific interfaces

**Comment by rgacogne** [comment]
> The biggest one is the fact that you are storing QNames in BPF Table like that (counter, qtype) -> qname

I believe you mean qname -> (counter, qtype)? The reason for that was to be able to block all types, using `ANY`, without having to do two lookups. Of course it means you can either block a type or all types, but not two types for the same name.

**Comment by MiniPierre** [comment]
Doesn't it creates problems when you want to block multiple qtypes for a same name, as the key already exists ? For example if A records are blocked for let's say localhost, you would not be able to block AAAA records, whereas you could do it by using (qname, qtype) keys

**Comment by rgacogne** [comment]
Yes, it does. At the time I decided that I preferred to have a way to be able to block all types for a name than having a way to block two different types, since I couldn't do two lookups without exceeding the maximum of instructions. This number has been raised since so perhaps we could afford to do two lookups now, one for ANY (since I expect most names to be blocked for all types) and one for the exact type.

**Comment by MiniPierre** [comment]
Hey, it took quite a long time but the filter is finally done. Now how should we proceed ? Should I do a pull request with the BPF source code only ?

**Comment by rgacogne** [comment]
Hello! Yes, a pull request would be a very good way to shared your work so we can discuss how to best integrate it :+1:

**Comment by MiniPierre** [comment]
Pull request is done, I'm sorry if it is not correctly formatted as it is my very first one

**Comment by rgacogne** [comment]
Closing since we merged #10498
